sw/host/opentitanlib/src/transport/dediprog/spi.rs

```diff
@@ -0,0 +1,367 @@
+use crate::io::spi::{
+    AssertChipSelect, ClockPolarity, SpiError, Target, TargetChipDeassert, Transfer, TransferMode,
+};
+use crate::transport::dediprog::{ClockSpeed, Command, Inner};
+
+pub struct DediprogSpi {
+    inner: Rc<RefCell<Inner>>,
+    max_chunk_size: usize,
+}
+
+impl DediprogSpi {
+    const MAX_GENERIC_DATA_LEN: usize = 16;
+
+    const READ_CHUNK_SIZE: usize = 512;
+    const WRITE_CHUNK_SIZE: usize = 256;
+
+    pub fn open(inner: Rc<RefCell<Inner>>) -> Result<Self> {
+        let this = Self {
+            inner,
+            max_chunk_size: 65535 * 256,
+        };
+        this.set_spi_clock()?;
+        Ok(this)
+    }
+
+    fn set_spi_clock(&self) -> Result<()> {
+        self.inner.borrow().device.write_control(
+            rusb::request_type(
+                rusb::Direction::Out,
+                rusb::RequestType::Vendor,
+                rusb::Recipient::Endpoint,
+            ),
+            Command::SetSpiClk as u8,
+            self.inner.borrow().spi_clock as u16,
+            0,
+            &[],
+        )?;
+        Ok(())
+    }
+
+    fn read_sfdp(&self, mut address: u32, mut rbuf: &mut [u8]) -> Result<()> {
+        // Read 16 bytes at a time, returning a single result.
+        while rbuf.len() > 0 {
+            let chunk_size = std::cmp::min(rbuf.len(), Self::MAX_GENERIC_DATA_LEN);
+            let addr_bytes = address.to_be_bytes();
+            let sub_cmd: [u8; 5] = [0x5a, addr_bytes[1], addr_bytes[2], addr_bytes[3], 0];
+            self.transmit(&sub_cmd, chunk_size)?;
+            self.receive(&mut rbuf[..chunk_size])?;
+            rbuf = &mut rbuf[chunk_size..];
+            address += chunk_size as u32;
+        }
+        Ok(())
+    }
+
+    fn transmit(&self, wbuf: &[u8], rbuf_len: usize) -> Result<()> {
+        self.inner.borrow().device.write_control(
+            rusb::request_type(
+                rusb::Direction::Out,
+                rusb::RequestType::Vendor,
+                rusb::Recipient::Endpoint,
+            ),
+            Command::Transceive as u8,
+            rbuf_len as u16,
+            0,
+            wbuf,
+        )?;
+        Ok(())
+    }
+
+    /// Receive data for a single SPI operation, using one or more USB packets.
+    fn receive(&self, rbuf: &mut [u8]) -> Result<()> {
+        self.inner.borrow().device.read_control(
+            rusb::request_type(
+                rusb::Direction::In,
+                rusb::RequestType::Vendor,
+                rusb::Recipient::Endpoint,
+            ),
+            Command::Transceive as u8,
+            0,
+            0,
+            rbuf,
+        )?;
+        Ok(())
+    }
+
+    fn eeprom_read_transaction(&self, cmd: &eeprom::Cmd, rbuf: &mut [u8]) -> Result<()> {
+        let mut usbcmd = [0u8; 10];
+        match (
+            cmd.get_opcode(),
+            cmd.get_address_len(),
+            cmd.get_dummy_cycles(),
+        ) {
+            ([0x5A], 3, 8) => {
```

Comment:
Use symbolic names for eeprom opcodes here (e.g. `SpiFlash::READ_SFDP`)

Reply:
Done